### PR TITLE
Corrects Error Codes from cudaHostRegister

### PR DESCRIPTION
**Pull request by LucasLLC**
Causing some terrible error messages e.g. :

```
# printing directly: cudaError.???
# casting to int first: 712

Traceback (most recent call last):
  File "/data/users/lpasqualin/fbsource/fbcode/scripts/lpasqualin/playground.py", line 15, in <module>
    main()
  File "/data/users/lpasqualin/fbsource/fbcode/scripts/lpasqualin/playground.py", line 11, in main
    _create_cpu_state_dict(sd, share_memory=True, pin_memory=True)
  File "/home/lpasqualin/pytorch/torch/distributed/_state_dict_utils.py", line 436, in _create_cpu_state_dict
    ret = _iterate_state_dict(
          ^^^^^^^^^^^^^^^^^^^^
  File "/home/lpasqualin/pytorch/torch/distributed/_state_dict_utils.py", line 143, in _iterate_state_dict
    ret = {
          ^
  File "/home/lpasqualin/pytorch/torch/distributed/_state_dict_utils.py", line 144, in <dictcomp>
    key: _iterate_state_dict(
         ^^^^^^^^^^^^^^^^^^^^
  File "/home/lpasqualin/pytorch/torch/distributed/_state_dict_utils.py", line 125, in _iterate_state_dict
    ret = tensor_func(iter_object, pg, device, companion_obj)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lpasqualin/pytorch/torch/distributed/_state_dict_utils.py", line 428, in tensor_func
    succ == 0
AssertionError: Pinning shared memory failed with error-code: cudaError.???
```

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/distributed/_state_dict_utils.py [hunk 1]
```diff
@@ -411,17 +411,17 @@ def tensor_func(
             if pin_memory:
 
                 def unpin_memory(t):
-                    succ = torch.cuda.cudart().cudaHostUnregister(t.data_ptr())
+                    succ = int(torch.cuda.cudart().cudaHostUnregister(t.data_ptr()))
                     assert (
```
Comment (rsdcastro) [inline torch/distributed/_state_dict_utils.py:415]: I'm curious to know whether torch.cuda.check_error would be an option to check for errors here.

Comment (LucasLLC) [reply torch/distributed/_state_dict_utils.py:415]: Good catch! I think check error would be reasonable but it currently doesn't support attaching metadata (the assertion string I think is useful).

Maybe we can circle back and extend `check_error`

File: torch/distributed/_state_dict_utils.py [hunk 2]
```diff
@@ -411,16 +411,18 @@ def tensor_func(
             if pin_memory:
 
                 def unpin_memory(t):
-                    succ = torch.cuda.cudart().cudaHostUnregister(t.data_ptr())
+                    succ = int(torch.cuda.cudart().cudaHostUnregister(t.data_ptr()))
```
Comment (wconstab) [inline torch/distributed/_state_dict_utils.py:414]: what is the type returned by cudaHostUnregister?

if it's not always an int, i worried about this raising an exception

Comment (LucasLLC) [reply torch/distributed/_state_dict_utils.py:414]: Generally uses this native enum: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1g3f51e3575c2178246db0a94a430e0038

which maps to int values that are all < 10k

**Pull request closed without merging**